### PR TITLE
Adding multi-tenancy support to MongoDb Panache throught dynamic database selection

### PR DESCRIPTION
**Pull request by pedroh-pereira**
Based on #14789

It consists in a interface called MongoDatabaseResolver that can be implemented to get the database name to be used on mongo operations execution
So if I have a MongoDatabaseResolver implementation the database to be used in the current mongo operation will be the return of resolve() method, but the resolver will be successful only for the entities that doesn't specify the database through the @MongoEntity(database="")

Disclaimer:
After this PR be approved I can help to provide a detailed documentation

@loicmathieu  @evanchooly

### CODE REVIEW

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java [hunk 1]
```diff
@@ -74,7 +76,7 @@ public Uni<Void> persist(Iterable<?> entities) {
                 objects.add(entity);
             }
 
-            if (objects.size() > 0) {
+            if (!objects.isEmpty()) {
```
Comment (evanchooly) [inline extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java:79]: There are a ton of unrelated changes to this PR.  While I applaud code clean up, it does leave a fair bit to scroll through to get to relevant details.

File: integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/ResolversPropertiesConstants.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package io.quarkus.it.mongodb.panache;
+
+public class ResolversPropertiesConstants {
+    public static final String MONGO_DATABASE_RESOLVER_ENABLED_PROPERTY = "quarkus.mongodb.database-resolver.enabled";
```
Comment (evanchooly) [inline integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/ResolversPropertiesConstants.java:4]: Adding a class just to hold a single constant seems less than ideal.

File: integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/profiles/MongoDatabaseResolverTestProfile.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package io.quarkus.it.mongodb.panache.profiles;
+
+import io.quarkus.test.junit.QuarkusTestProfile;
+
+public class MongoDatabaseResolverTestProfile implements QuarkusTestProfile {
```
Comment (evanchooly) [inline integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/profiles/MongoDatabaseResolverTestProfile.java:5]: Does this serve an actual purpose since it doesn't override anything from the parent type?

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/MongoDatabaseResolver.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package io.quarkus.mongodb.panache.common;
+
+/*
+ * This interface can be used to resolve the mongo database name in runtime, it'll help to implement multi-tenancy using tenant per database approach
+ */
+public interface MongoDatabaseResolver {
+    String resolve();
```
Comment (evanchooly) [inline extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/MongoDatabaseResolver.java:7]: What still eludes me is how this is actually going to help resolve tenancy.  Without some context, there's not much to be done here.  Is the idea that this resolver would read a different config file?  How are multiple tenants supported by this?

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/BeanUtils.java [hunk 5]
```diff
@@ -82,4 +92,9 @@ public static String getDatabaseName(MongoEntity mongoEntity, String clientBeanN
                 "The database attribute was not set for the @MongoEntity annotation neither was the database property configured for the named Mongo Client (via 'quarkus.mongodb.%s.database')",
                 mongoEntity.clientName()));
     }
+
+    public static Optional<String> getDatabaseNameFromResolver() {
```
Comment (evanchooly) [inline extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/BeanUtils.java:96]: This can probably just be one method combined with `getValueFromArcInstance` since no one else is calling that.

File: integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/resolvers/ResolverEntityWithDatabase.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package io.quarkus.it.mongodb.panache.resolvers;
+
+import io.quarkus.mongodb.panache.common.MongoEntity;
+
+public class ResolverEntityWithDatabase {
+    public static final String DATABASE_NAME = "database-from-anotation";
+
+    @MongoEntity(database = DATABASE_NAME)
+    public static class Entity {
```
Comment (evanchooly) [inline integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/resolvers/ResolverEntityWithDatabase.java:9]: Why isn't this just a top level class?  I know this is "just" test code but there doesn't seem to be a reason for the nesting here.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java (hunk 1 above)
Comment (pedroh-pereira) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java:79]: Sorry about that, I can undo these changes if needed.

File: integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/ResolversPropertiesConstants.java (hunk 2 above)
Comment (pedroh-pereira) [reply integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/ResolversPropertiesConstants.java:4]: The initial idea was submit more than one resolver in this branch and I has implemented three resolvers and I decided to submit just one per PR and I was using this one to store the constants of these other, but really it's not appropriate to this PR and I can fix it.

File: integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/profiles/MongoDatabaseResolverTestProfile.java (hunk 3 above)
Comment (pedroh-pereira) [reply integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/profiles/MongoDatabaseResolverTestProfile.java:5]: It's a mistake, I was not using it anymore. The initial idea was mapping some properties to be used for the test using this profiles. I'll remove it

File: integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/ResolversPropertiesConstants.java (hunk 2 above)
Comment (evanchooly) [reply integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/ResolversPropertiesConstants.java:4]: I would really like to see that in action, honestly. I'm having trouble seeing how this solves the problem.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/MongoDatabaseResolver.java (hunk 4 above)
Comment (pedroh-pereira) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/MongoDatabaseResolver.java:7]: For example I would implement it like:
In my API I receive a request header 'x-customer-id' identifying the customer, for each request I'll get the database and put into a ThreadLocal, so the MongoDatabaseResolver could be like:
`@ApplicationScoped
public class CustomMongoCollectionPrefixResolver implements MongoDatabaseResolver {

  @Override
  public String resolve() {
    return DatabaseContext.databaseName.get();
  }
  
}`
In this scenario the database would be selected dynamically

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/BeanUtils.java (hunk 5 above)
Comment (pedroh-pereira) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/BeanUtils.java:96]: The reason is the same of previous commen. The initial idea was submit more than one resolver and I removed the other ones that was using this common method. Currently, really doesn't make senses and I'll fix it

File: integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/resolvers/ResolverEntityWithDatabase.java (hunk 6 above)
Comment (pedroh-pereira) [reply integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/resolvers/ResolverEntityWithDatabase.java:9]: The idea is use the constant in test assertations to avoid passing a hard coded string in assertation, but If you thinks that it is not a worth reason I can fix it.

Comment (evanchooly) [reply integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/resolvers/ResolverEntityWithDatabase.java:9]: That constant can still be used without having to nest like that.

File: integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/ResolversPropertiesConstants.java (hunk 2 above)
Comment (pedroh-pereira) [reply integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/ResolversPropertiesConstants.java]: Currently, It is not beeing used and I'll remove it

Comment (pedroh-pereira) [reply integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/ResolversPropertiesConstants.java]: So the initial idea was use a different profile to each resolver, and the profile would use these class to map the properties.
For example:
Wondering if we have 2 resolvers MongoClientNameResolver and ongoDabaseResolver, both would have a const and a test profile.
The key point is the that I follow a different approach to use in the test, and this class doesn't need to be use anymore.

Comment (pedroh-pereira) [reply integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/ResolversPropertiesConstants.java]: I committed a change removing it

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/MongoDatabaseResolver.java (hunk 4 above)
Comment (evanchooly) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/MongoDatabaseResolver.java:7]: fair enough

File: integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/resolvers/BaseMongodbPanacheResolversTest.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
+package io.quarkus.it.mongodb.panache.resolvers;
```
Comment (evanchooly) [inline integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/resolvers/BaseMongodbPanacheResolversTest.java:1]: It doesn't look like there's a test case where no `Resolver` has been defined.  Since we're injecting that mock, it's always going to find a `Resolver`, yes?

**Comment by evanchooly** [comment]
There are a lot of unrelated commits attached to this PR.  Looks like you could use a rebase.

**Comment by pedroh-pereira** [comment]
> There are a lot of unrelated commits attached to this PR. Looks like you could use a rebase.

Fixed, I had a problem with my IDE

File: integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/resolvers/BaseMongodbPanacheResolversTest.java (hunk 7 above)
Comment (pedroh-pereira) [reply integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/resolvers/BaseMongodbPanacheResolversTest.java:1]: Yes, that`s it. This no `Resolver` case is already covered by the other tests since we`re not implementing it in the other test classes, but if needed I can create a test class only to cover this scenario/

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/MongoDatabaseResolver.java [hunk 8]
```diff
@@ -0,0 +1,8 @@
+package io.quarkus.mongodb.panache.common;
+
+/*
+ * This interface can be used to resolve the mongo database name in runtime, it'll help to implement multi-tenancy using tenant per database approach
```
Comment (gsmet) [inline extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/MongoDatabaseResolver.java:4]:
```suggestion
 * This interface can be used to resolve the MongoDB database name at runtime, it allows to implement multi-tenancy using a tenant per database approach.
```

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java (hunk 1 above)
Comment (gsmet) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java:79]: When doing such changes, it's recommended to do a separate commit (or even pull request) with the change. Given the changes are relatively safe, I think we can keep them as is.

**Review by gsmet — CHANGES_REQUESTED** [review]
Thanks for the pull request.

I added a suggestion for the Javadoc. Also, once you are done, please squash the commits into one commit.

I will let @evanchooly and @loicmathieu review the code though.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java (hunk 1 above)
Comment (pedroh-pereira) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java:79]: @gsmet sure, thanks for revision and the tip!

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/MongoDatabaseResolver.java (hunk 8 above)
Comment (pedroh-pereira) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/MongoDatabaseResolver.java]: Thanks for your suggestion, I've added it and squashed all commits into one.

File: integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/resolvers/BaseMongodbPanacheResolversTest.java (hunk 7 above)
Comment (evanchooly) [reply integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/resolvers/BaseMongodbPanacheResolversTest.java:1]: Duh.  You're right.  Nevermind me.  :)

**Review by evanchooly — APPROVED** [review]

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/MongoDatabaseResolver.java (hunk 4 above)
Comment (pedroh-pereira) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/MongoDatabaseResolver.java:7]: @evanchooly It's missing a new blank line at the end. Should I commit it again?

**Review by loicmathieu — CHANGES_REQUESTED** [review]
It misses the documentation part in the guides.

I think the test is not very accurate about a real use case. It mock the resolver and everything lies in the `test` directory.

We usually test new functionalities in a more "real world" manner.

For ex:

- Create a resolver that resvole the database name from a request header
- At startup, add "red" in a collection in database1 and "blue" in the same collection in database2
- Create a REST endpoint (or enhance an existing one) with an operation that get the red/blue from the database
- Test from restassured that when you pass the header "blue" you get "blue" from the database

This will also be a good example for the guide.

**Comment by pedroh-pereira** [comment]
> It misses the documentation part in the guides.
> 
> I think the test is not very accurate about a real use case. It mock the resolver and everything lies in the `test` directory.
> 
> We usually test new functionalities in a more "real world" manner.
> 
> For ex:
> 
> * Create a resolver that resvole the database name from a request header
> * At startup, add "red" in a collection in database1 and "blue" in the same collection in database2
> * Create a REST endpoint (or enhance an existing one) with an operation that get the red/blue from the database
> * Test from restassured that when you pass the header "blue" you get "blue" from the database
> 
> This will also be a good example for the guide.

I’m working on it, probably until tomorrow I will override these commits with all changes

**Comment by pedroh-pereira** [comment]
> It misses the documentation part in the guides.
> 
> I think the test is not very accurate about a real use case. It mock the resolver and everything lies in the `test` directory.
> 
> We usually test new functionalities in a more "real world" manner.
> 
> For ex:
> 
> * Create a resolver that resvole the database name from a request header
> * At startup, add "red" in a collection in database1 and "blue" in the same collection in database2
> * Create a REST endpoint (or enhance an existing one) with an operation that get the red/blue from the database
> * Test from restassured that when you pass the header "blue" you get "blue" from the database
> 
> This will also be a good example for the guide.

@loicmathieu @evanchooly  Done
